### PR TITLE
Initial lightweight serialization && deserialization for endorsements…

### PR DESCRIPTION
**Pull request by sydhds**
… in a block header

### CODE REVIEW

File: massa-models/src/block.rs [hunk 1]
```diff
@@ -357,6 +359,117 @@ impl Deserializer<Block> for BlockDeserializer {
     }
 }
 
+/// Deserializer for `Block`
+pub struct BlockDeserializerLW {
```
Comment (damip) [inline massa-models/src/block.rs:363]: what's "LW" ?

File: massa-factory-worker/src/endorsement_factory.rs [hunk 2]
```diff
@@ -187,7 +187,7 @@ impl EndorsementFactoryWorker {
                     index: index as u32,
                     endorsed_block,
                 },
-                EndorsementSerializer::new(), // TODO reuse self.endorsement_serializer
```
Comment (AurelienFT) [inline massa-factory-worker/src/endorsement_factory.rs:190]: I think this todo should still be here. The goal is too avoid initializing a new serializer each loop run

File: massa-models/src/endorsement.rs [hunk 3]
```diff
@@ -219,6 +219,108 @@ impl Deserializer<Endorsement> for EndorsementDeserializer {
...
+pub struct EndorsementSerializerLW {
+    // slot_serializer: SlotSerializer,
+    u32_serializer: U32VarIntSerializer,
+}
+
+impl EndorsementSerializerLW {
+    /// Creates a new `EndorsementSerializerLW`
+    pub fn new() -> Self {
+        EndorsementSerializerLW {
+            // slot_serializer: SlotSerializer::new(),
```
Comment (AurelienFT) [inline massa-models/src/endorsement.rs:233]: I think this is outdated code.

File: massa-models/src/endorsement.rs [hunk 4]
```diff
@@ -219,6 +219,108 @@ impl Deserializer<Endorsement> for EndorsementDeserializer {
...
+    /// EndorsementSerializerLW::new().serialize(&endorsement, &mut buffer).unwrap();
+    /// let (rest, deserialized) = EndorsementDeserializerLW::new(10).deserialize::<DeserializeError>(&buffer).unwrap();
+    /// assert_eq!(rest.len(), 0);
+    /// assert_eq!(deserialized.index, endorsement.index);
+    /// ```
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], Endorsement, E> {
+        println!("[EndorsementDeserializerLW] start");
```
Comment (AurelienFT) [inline massa-models/src/endorsement.rs:307]: Debug print

File: massa-models/src/block.rs [hunk 5]
```diff
@@ -357,6 +358,117 @@ impl Deserializer<Block> for BlockDeserializer {
...
```
Comment (AurelienFT) [inline massa-models/src/block.rs:362]: I don't understand why you have implemented this which is the same code as BlockDeserializer just use the lightweight deserializer for header. Why not using it directly on block deserializer ?

**Review by AurelienFT — COMMENTED** [review]
Very good work ! I just don't understand why you don't override the existing deserializers for blocks and block header instead of creating LW mode.

File: massa-models/src/endorsement.rs (hunk 3 above)
Comment (sydhds) [reply massa-models/src/endorsement.rs:233]: removed

File: massa-models/src/endorsement.rs (hunk 4 above)
Comment (sydhds) [reply massa-models/src/endorsement.rs:307]: Removed

File: massa-factory-worker/src/endorsement_factory.rs (hunk 2 above)
Comment (sydhds) [reply massa-factory-worker/src/endorsement_factory.rs:190]: I've added it again. Maybe I will adress the issue in this PR too...

Comment (sydhds) [reply massa-factory-worker/src/endorsement_factory.rs:190]: Field added to EndorsementFactoryWorker

**Review by AurelienFT — COMMENTED** [review]
Sorry I have forgotten a part on my previous comment: I think you could use only one block header serializer also instead of having two versions.

**Comment by sydhds** [comment]
> Sorry I have forgotten a part on my previous comment: I think you could use only one block header serializer also instead of having two versions.

Done

**Review by AurelienFT — APPROVED** [review]

File: massa-bootstrap/src/tests/tools.rs [hunk 6]
```diff
@@ -373,7 +373,7 @@ pub fn get_boot_state() -> BootstrapableGraph {
                                 index: 1,
                                 endorsed_block: get_dummy_block_id("p1"),
                             },
-                            EndorsementSerializer::new(),
+                            EndorsementSerializerLW::new(),
```
Comment (AurelienFT) [inline massa-bootstrap/src/tests/tools.rs:376]: Shouldn't be LW here

File: massa-bootstrap/src/tests/tools.rs [hunk 7]
```diff
@@ -383,7 +383,7 @@ pub fn get_boot_state() -> BootstrapableGraph {
                                 index: 3,
                                 endorsed_block: get_dummy_block_id("p1"),
                             },
-                            EndorsementSerializer::new(),
+                            EndorsementSerializerLW::new(),
```
Comment (AurelienFT) [inline massa-bootstrap/src/tests/tools.rs:386]: Shouldn't be LW here

File: massa-consensus-worker/src/tests/tools.rs [hunk 8]
```diff
@@ -491,7 +491,7 @@ pub fn create_endorsement(
         index,
         endorsed_block,
     };
-    Endorsement::new_wrapped(content, EndorsementSerializer::new(), sender_keypair).unwrap()
+    Endorsement::new_wrapped(content, EndorsementSerializerLW::new(), sender_keypair).unwrap()
```
Comment (AurelienFT) [inline massa-consensus-worker/src/tests/tools.rs:494]: Shouldn't be LW her

File: massa-models/src/block.rs [hunk 9]
```diff
@@ -290,7 +292,7 @@ impl Deserializer<Block> for BlockDeserializer {
     ///                     index: 1,
     ///                     endorsed_block: BlockId(Hash::compute_from("blk1".as_bytes())),
     ///                 },
-    ///                 EndorsementSerializer::new(),
+    ///                 EndorsementSerializerLW::new(),
```
Comment (AurelienFT) [inline massa-models/src/block.rs:295]: Shouldn't be LW here

File: massa-models/src/block.rs [hunk 10]
```diff
@@ -300,7 +302,7 @@ impl Deserializer<Block> for BlockDeserializer {
     ///                     index: 3,
     ///                     endorsed_block: BlockId(Hash::compute_from("blk2".as_bytes())),
     ///                 },
-    ///                 EndorsementSerializer::new(),
+    ///                 EndorsementSerializerLW::new(),
```
Comment (AurelienFT) [inline massa-models/src/block.rs:305]: Shouldn't be LW here

File: massa-models/src/block.rs [hunk 11]
```diff
@@ -781,6 +812,140 @@ mod test {
...
+
+        // create block header
+        let orig_header = BlockHeader::new_wrapped(
+            BlockHeader {
+                slot: Slot::new(1, 1),
+                parents,
+                operation_merkle_root: Hash::compute_from("mno".as_bytes()),
+                endorsements: vec![Endorsement::new_wrapped(
+                    endorsement,
+                    EndorsementSerializerLW::new(),
```
Comment (AurelienFT) [inline massa-models/src/block.rs:919]: Shouldn't be LW here

File: massa-models/src/endorsement.rs [hunk 12]
```diff
@@ -219,6 +220,106 @@ impl Deserializer<Endorsement> for EndorsementDeserializer {
...
+    /// let endorsement = Endorsement {
+    ///   slot: Slot::new(1, 2),
+    ///   index: 0,
+    ///   endorsed_block: BlockId(Hash::compute_from("test".as_bytes()))
+    /// };
+    /// let mut buffer = Vec::new();
+    /// EndorsementSerializerLW::new().serialize(&endorsement, &mut buffer).unwrap();
+    /// ```
+    fn serialize(&self, value: &Endorsement, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
+        // self.slot_serializer.serialize(&value.slot, buffer)?;
```
Comment (AurelienFT) [inline massa-models/src/endorsement.rs:261]: Dead code.

File: massa-protocol-exports/src/tests/tools.rs [hunk 13]
```diff
@@ -231,7 +231,7 @@ pub fn create_endorsement() -> WrappedEndorsement {
         index: 0,
         endorsed_block: BlockId(Hash::compute_from(&[])),
     };
-    Endorsement::new_wrapped(content, EndorsementSerializer::new(), &keypair).unwrap()
+    Endorsement::new_wrapped(content, EndorsementSerializerLW::new(), &keypair).unwrap()
```
Comment (AurelienFT) [inline massa-protocol-exports/src/tests/tools.rs:234]: Shouldn't be LW here

File: massa-bootstrap/src/tests/tools.rs (hunk 6 above)
Comment (sydhds) [reply massa-bootstrap/src/tests/tools.rs:376]: done

File: massa-bootstrap/src/tests/tools.rs (hunk 7 above)
Comment (sydhds) [reply massa-bootstrap/src/tests/tools.rs:386]: done

File: massa-consensus-worker/src/tests/tools.rs (hunk 8 above)
Comment (sydhds) [reply massa-consensus-worker/src/tests/tools.rs:494]: done

File: massa-models/src/block.rs (hunk 9 above)
Comment (sydhds) [reply massa-models/src/block.rs:295]: done

File: massa-models/src/block.rs (hunk 11 above)
Comment (sydhds) [reply massa-models/src/block.rs:919]: done

File: massa-models/src/block.rs (hunk 10 above)
Comment (sydhds) [reply massa-models/src/block.rs:305]: done

File: massa-models/src/endorsement.rs (hunk 12 above)
Comment (sydhds) [reply massa-models/src/endorsement.rs:261]: removed

File: massa-protocol-exports/src/tests/tools.rs (hunk 13 above)
Comment (sydhds) [reply massa-protocol-exports/src/tests/tools.rs:234]: done

**Pull request merged**